### PR TITLE
WIP: Fix netplan systemd dbus race

### PR DESCRIPTION
**Pull request by holmanb**
## Additional Context
This is a rough sketch of how we might fix the persistent dbus/networkd/netplan race that we've been seeing for some time now. The commit messages and comments explain the code changes.

Feel free to bikeshed implementation details.

Fixes https://github.com/canonical/cloud-init/issues/4044

TODO: Clean up logic
TODO: Tests

Review by an SRU reviewer desired after details are hammered out, but before merge.

## Merge type
- [ ] Squash merge using "Proposed Commit Message"
- [x] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: cloudinit/cmd/main.py [hunk 1]
```diff
@@ -431,8 +431,24 @@ def main_init(name, args):
...
+        # cloud-init.service is started before dbus.socket, but does not block
+        # it. Retry until dbus becomes available.
+        while not init.apply_network_config(bring_up=bring_up_interfaces):
+            LOG.warning(
+                "Dbus is not available to notify systemd-networkd of"
+                "a new configuration from netplan. Retrying."
+            )
+            # Why retry every 0.042? it uses a fraction of a percent of one CPU
+            # This seems like a reasonable tradeoff between "get there fast"
+            # and "leave lots of resources available for others".
```
Comment (cjp256) [inline cloudinit/cmd/main.py:450]: lol this made me laugh, but really I want to know _why 0.042_ :D

Comment (holmanb) [reply cloudinit/cmd/main.py:450]: Heh, I'm a big fan of a certain universal guide ;)

Comment (holmanb) [reply cloudinit/cmd/main.py:450]: Hitchhiker's reference aside, I can run something like:
```python
>>> import time
>>> while 1:
...     time.sleep(N)
```

And it doesn't even register in `htop` above around `N=0.02`. With that in mind, somewhere between `0.02` and `0.1` seemed acceptable for a WIP POC like this - happy to make it whatever seems best :D

File: cloudinit/net/activators.py [hunk 2]
```diff
@@ -17,12 +17,14 @@ class NoActivatorException(Exception):
...
+def _alter_interface(cmd, device_name, fail_on_stderr=False) -> bool:
     LOG.debug("Attempting command %s for device %s", cmd, device_name)
     try:
-        (_out, err) = subp.subp(cmd)
-        if len(err):
+        err = subp.subp(cmd).stderr
+        if err and fail_on_stderr and "Failed to connect system bus:" in err:
             LOG.warning("Running %s resulted in stderr output: %s", cmd, err)
+            return False
+        LOG.info("Running %s resulted in stderr output: %s", cmd, err)
```
Comment (TheRealFalcon) [inline cloudinit/net/activators.py:27]: I don't think we want this log when there is no stderr.

File: cloudinit/net/activators.py [hunk 3]
```diff
@@ -17,12 +17,13 @@ class NoActivatorException(Exception):
...
 
 
-def _alter_interface(cmd, device_name) -> bool:
+def _alter_interface(cmd, device_name, fail_on_stderr=False) -> bool:
     LOG.debug("Attempting command %s for device %s", cmd, device_name)
     try:
-        (_out, err) = subp.subp(cmd)
-        if len(err):
+        err = subp.subp(cmd).stderr
+        if err and fail_on_stderr and "Failed to connect system bus:" in err:
```
Comment (TheRealFalcon) [inline cloudinit/net/activators.py:24]: I might be a little pedantic here, but don't we want the LOG anytime we have stderr? Something like:
```python
try:
    err = subp.subp(cmd).stderr
    if err:
        LOG.warning("...")
        if fail_on_stderr and "Failed to connect system bus:" in err:
            return False
    ...
```
?

**Comment by slyon** [comment]
This feels like a workaround to me, and I feel we should rather try to fix the original ordering problem.

I do not fully understand the cloud-init ordering stages we're interacting with here. Maybe you can provide some more context?

In the commit message you suggest this ordering:
systemd-netword.service < systemd-networkd-network-wait-online.service < cloud-init.service

But cloud-init.service also needs dbus.socket, therefore implementing some polling to wait for it. Why can't we just define a systemd dependency, so that cloud-init.service runs AFTER systemd-networkd-wait-online.service AND dbus.socket?

Or alternatively have an additional cloud-init stage, that runs BEFORE systemd-networkd.service, only calling `netplan generate` instead of `netplan apply` (so doesn't need DBus and will be applied by systemd-networkd starting afterwards, implicitly). IIUC there is already a previous cloud-init stage where you spin up some ephemeral DHCP connection, using a DHCP client manually, in order to fetch the actual Netplan configuration from an online data-source. So the Netplan YAML config should be available before starting systemd-networkd, right?

**Pull request closed without merging**